[python-package] Correctly recognize LGBMClassifier(num_class=2, objective="multiclass") as multiclass classification

RektPunk opened the pull request:
fixes #6519
fixes #3636

I created a pull request addressing issue [#6519](https://github.com/microsoft/LightGBM/issues/6519), which I've also encountered. 
I modified the code to use `np.vstack` only when `result.shape` look like `(num_data,)`, and some typos.

jameslamb reviewed (changes requested):
thanks, could you please add a unit test covering this behavior, to be sure it isn't broken by future refactorings?

RektPunk commented:
@microsoft-github-policy-service agree

jameslamb commented on tests/python_package_test/test_sklearn.py:
```diff
@@ -719,6 +719,21 @@ def test_predict():
         np.testing.assert_allclose(res_engine, res_sklearn_params)
 
+    # Test multiclass binary classification
+    num_samples, num_classes = 20, 2
+    rng = np.random.Generator(np.random.PCG64())
+    X_train = rng.uniform(low=0, high=1, size=[num_samples, 3])
+    y_train = rng.choice([0, 1], size=num_samples)
```
The randomness here could sometimes lead to the test failing. For example, it's possible for `y_train` to contain all 0s, which would result in an error at training time... not too unlikely, given the small data sizes involved here.

I directly experienced that trying to run the reproducible example from #6519.

Could you please remove randomness here and just create these arrays with `np.array` literals or something else deterministic?

RektPunk replied:
I remove randomness in https://github.com/microsoft/LightGBM/pull/6524/commits/66b372ae5d5dcdedbf9862fa877d2e65af19382b.

jameslamb reviewed (changes requested):
Thanks very much.

#6519 is reporting the same root cause as #3636, and the solution you've arrived at here looks very similar to what @jmoralez recommended there (https://github.com/microsoft/LightGBM/issues/3636#issuecomment-1376684384). There he recommended using `result.ndim` instead of `len(result.shape)`... I don't have a strong opinion on one of those being preferable to another. I think the `len(result.shape)` approach you've taken is fine.

Re-reading #3636 though, I think there is one other place where `self._n_classes > 2` is used to decide whether binary of multiclass classification is being performed, and which needs to be changed.

@rabyj mentioned something related to metrics at https://github.com/microsoft/LightGBM/issues/3636#issuecomment-1326851476, which I think is about this behavior where `LGBMClassifier` rewrites binary classification metrics to multiclass metrics and vice-versa:

https://github.com/microsoft/LightGBM/blob/a5054f7752385f26a4fc0a100475ad0870483ba9/python-package/lightgbm/sklearn.py#L1254

This is making me think that maybe `LGBMClassifier` should stop relying on `self._n_classes > 2` as a proxy for "is doing multiclass classification".

What do you think about adding a private property on `LGBMClassifier` that has something like this?

```python
@property
def __is_multiclass(self) -> bool:
   multiclass_objectives = {"multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr"} 
   return (
        self._n_classes > 2
        or isinstance(self._objective, str) and self._objective in multiclass_objectives
    )
```

And then using that in place of these `self._n_classes > 2` conditions, where appropriate, e.g.

```python
if self.__is_multiclass:
   # do multiclass classification things
else:
   # do binary classification things
```

If you agree, then please also add a case within this test checking that the behavior is correct when using `num_class=2, objective="multiclass"`:

https://github.com/microsoft/LightGBM/blob/a5054f7752385f26a4fc0a100475ad0870483ba9/tests/python_package_test/test_sklearn.py#L781

---

Sorry this is a lot... there is a lot of indirection in the `scikit-learn` estimator, some of it imposed by `scikit-learn` itself, so seemingly simple things can become complicated fast 😅 

Other references you might find relevant:

* #6002
* #5675

RektPunk commented on tests/python_package_test/test_sklearn.py:
```diff
@@ -719,6 +719,18 @@ def test_predict():
+    clf = lgb.LGBMClassifier(objective="multiclass", num_classes=2).fit(X_train, y_train)
+
+    res_engine = gbm.predict(X_train)
+    res_sklearn = clf.predict_proba(X_train)
```
I add test for `LGBMClassifier.predict()` in https://github.com/microsoft/LightGBM/pull/6524/commits/66b372ae5d5dcdedbf9862fa877d2e65af19382b.

RektPunk commented:
Thanks for review, @jameslamb. As you mentioned, I added `__is_multiclass` and also added tests for the eval metric (if the test you had in mind is correct 😅 )

jameslamb commented on tests/python_package_test/test_sklearn.py:
```diff
@@ -719,6 +719,24 @@ def test_predict():
         np.testing.assert_allclose(res_engine, res_sklearn_params)
 
+    # Test multiclass binary classification
+    num_samples, num_classes = 5, 2
+    X_train = np.ones((num_samples, 3))
+    y_train = np.zeros((num_samples,))
```
```suggestion
    num_samples = 100
    num_classes = 2
    X_train = np.linspace(start=0, stop=10, num=num_samples*3).reshape(num_samples, 3)
    y_train = np.repeat([0, 1], repeats=num_samples/2)
```

It's a point of minor personal style preference, but could you please avoid using one line to assign multiple constants?

And could you please follow this suggestion to use 100 samples instead of 5, and a target that actually has samples from 2 classes, so some minimal training actually happens?

RektPunk replied:
Thanks, fix it as you suggested in bbbd5ba and 57e509c

jameslamb reviewed (changes requested):
New tests look great, thanks very much for your attention to detail there making them minimal and similar to the other existing tests! And for fixing typos in comments, much appreciated.

I just have a few more small suggestions.

RektPunk commented on python-package/lightgbm/sklearn.py:
```diff
@@ -1389,6 +1389,12 @@ def n_classes_(self) -> int:
+    @property
+    def __is_multiclass(self) -> bool:
+        """:obj:`bool`:  Indicator of whether the classifier is used for multiclass."""
+        multiclass_objectives = ("multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr")
```
Thanks, fix it in [6f1da7f](https://github.com/microsoft/LightGBM/pull/6524/commits/6f1da7f3a4be2d6b644cc5e2b63d0ab3a7e46be5)

jameslamb commented on python-package/lightgbm/sklearn.py:
```diff
@@ -10,6 +10,7 @@
 import scipy.sparse
 
 from .basic import (
+    _MULTICLASS_OBJECTIVES,
```
Could you put this in alphabetical order in this list? That'd be after `_LGBM_WeightType` below.

*(I'm not sure why `isort` didn't just do that automatically / the linter isn't complaining about that... sorry about that, I know it's no fun getting review comments like this)*

RektPunk replied:
Thank you for the review!. Because `isort` automatically sorts, it seems that this suggestion cannot be reflected 🥲.

jameslamb replied:
ah interesting, I wonder why `isort` chose to put this there 🤔 

well, not too important either way. Thanks for looking into it, and sorry for the annoying request. You've been very patient with me, I appreciate it!

RektPunk replied:
No worries at all! your requests are never annoying. Thank you for your kind words! 👏

jameslamb reviewed (approved):
Thanks very much! The changes and tests look great to me.

I left one more very very very minor suggestion.

Don't worry about the failing CI job (https://github.com/microsoft/LightGBM/actions/runs/9852647887/job/27218413942?pr=6524). I think that's the result of some recent change in the `numpy` or `pandas` nightlies, not anything done in this PR.

@jmoralez could you also review? I'm not 100% confident I've thought of all the implications of this change.

jameslamb commented:
Thanks very much for the work @RektPunk and for the help reviewing @jmoralez !

Pretty fund when one merge closes 2 issues 😁
